services/httpd/handler.go

```diff
@@ -993,8 +1017,437 @@ func fixLiterals(node influxql.Node) {
+			h.httpError(w, fmt.Sprintf("list buckets parameter is not an integer - offset=%q: %s", offsetStr, err.Error()), http.StatusBadRequest)
+			return
+		} else {
+			offset = int(o)
+		}
+	}
+
+	dbIndex := 0
+	rpIndex := 0
+	dbs := h.MetaClient.Databases()
+	if after != "" {
+		if dbIndex, rpIndex = findBucketIndex(dbs, db, rp); dbIndex < 0 {
+			h.httpError(w, fmt.Sprintf("list buckets \"after\" parameter not found: %q", after), http.StatusNotFound)
+			return
+		}
+	} else if offset > 0 {
+		if dbIndex, rpIndex = findBucketOffsetIndex(dbs, offset); dbIndex < 0 {
+			sendBuckets(w, []Bucket{})
+			return
+		}
+	}
+	sliceCap := limit
+	if sliceCap > (2 * defaultLimit) {
+		sliceCap = 2 * defaultLimit
+	}
+	buckets := make([]Bucket, 0, sliceCap)
+
+outer:
+	for ; dbIndex < len(dbs); dbIndex++ {
+		dbi := dbs[dbIndex]
+		if h.Config.AuthEnabled {
+			if err := h.QueryAuthorizer.AuthorizeDatabase(user, influxql.ReadPrivilege, dbi.Name); err != nil {
+				continue
+			}
+		}
+		for ; rpIndex < len(dbi.RetentionPolicies); rpIndex++ {
+			if limit > 0 {
+				buckets = append(buckets, *makeBucket(&dbi.RetentionPolicies[rpIndex], dbi.Name))
+				limit--
+			} else {
+				break outer
+			}
+		}
+		rpIndex = 0
+	}
+	sendBuckets(w, buckets)
+}
+
+func (h *Handler) oneBucket(w http.ResponseWriter, name string) *Bucket {
+	db, rp, err := bucket2dbrp(name)
+	if err != nil {
+		h.httpError(w, fmt.Sprintf("bucket %q: %s", name, err.Error()), http.StatusNotFound)
+		return nil
+	} else if dbi := h.MetaClient.Database(db); dbi == nil {
+		h.httpError(w, fmt.Sprintf("bucket not found: %q", name), http.StatusNotFound)
+		return nil
+	} else if rpi := dbi.RetentionPolicy(rp); rpi == nil {
+		h.httpError(w, fmt.Sprintf("bucket not found: %q", name), http.StatusNotFound)
+		return nil
+	} else {
+		return makeBucket(rpi, db)
+	}
+}
+
+func findBucketOffsetIndex(dbs []meta.DatabaseInfo, offset int) (dbIndex int, rpIndex int) {
+	for i, dbi := range dbs {
+		if offset > len(dbi.RetentionPolicies) {
+			offset -= len(dbi.RetentionPolicies)
+		} else {
+			dbIndex = i
+			rpIndex = offset
+			return dbIndex, rpIndex
+		}
+	}
+	return -1, -1
+}
+
+func findBucketIndex(dbs []meta.DatabaseInfo, db string, rp string) (dbIndex int, rpIndex int) {
+	for di, dbi := range dbs {
+		if dbi.Name != db {
+			continue
+		} else {
+			for ri, rpi := range dbi.RetentionPolicies {
+				if rpi.Name == rp {
+					if ri < (len(dbi.RetentionPolicies) - 1) {
+						return di, ri + 1
+					} else {
+						return di + 1, 0
+					}
+				}
+			}
+		}
+	}
+	return -1, -1
+}
+
+func sendBuckets(w http.ResponseWriter, buckets []Bucket) {
+	b, err := json.Marshal(buckets)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("list buckets marshaling error: %s", err.Error()), http.StatusBadRequest)
+		return
+	}
+	if _, err := w.Write(b); err != nil {
+		http.Error(w, fmt.Sprintf("list buckets error writing response: %s", err.Error()), http.StatusBadRequest)
```

Comment:
I think we should just log this - the first `Write` sends an implicit `StatusOk` if no other status is set first. In any case trying to write an error message over a connection that is failing to send data properly is likely doomed.

Reply:
'Hopefully' the client will detect the write error at the TCP layer.

Reply:
Fixed